File: TaskOne/src/test/java/com/ibs/litmus/controller/RegisterControllerTest.java
import org.mockito.MockitoAnnotations;
import org.springframework.web.servlet.ModelAndView;
import org.mockito.junit.jupiter.MockitoExtension;

import com.ibs.litmus.model.Person;
import com.ibs.litmus.myexceptions.PasswordException;
import com.ibs.litmus.repository.PersonRepo;

@ExtendWith(MockitoExtension.class)//annotn instd of openMocks
class RegisterControllerTest {
	@InjectMocks
	RegisterController rc;
	@Mock
    	PersonRepo repo;
	Person p ;
	@BeforeEach
	public void setup() {
		//MockitoAnnotations.initMocks(this);//jupiter 2-deprecated in jupiter 3 use openMocks or annotn
		//MockitoAnnotations.openMocks(this);
		p = new Person("testUserName", 50, "testName", "1971", "male", "password");
		System.out.println("Inside setup");
	}
	@AfterEach
	public void cleanup() {
		System.out.println("Inside cleanup");
	}
	@Test
	@DisplayName("save to db-pw lenth>6")
	void passwordTest1() throws PasswordException {
		p.setPassword("password");
		doReturn(p).when(repo).save(any(Person.class));
		rc.details(p);
		//ModelAndView mv = rc.details(p);
		assertEquals(8, p.getPassword().length(), "password length is greater than 6::criteria satisfied");
	}
	@Test
	@DisplayName("save to db-pw length=6")
	void passwordTest2() throws PasswordException {
		p.setPassword("passwo");
		doReturn(p).when(repo).save(any(Person.class));
		rc.details(p);
		assertEquals(6, p.getPassword().length(), "password length is 6::criteria satisfied");
	}
	@Test
	@DisplayName("dnt save to db,throw exceptn-pw length<6")
	void passwordTest3() {
		p.setPassword("pass");
		assertThrows(Exception.class, () -> rc.details(p),
				"password length is less than 6::criteria violation,throws exception");
	}
}
/*
class RegisterControllerTest {
	Person person=new Person("testUserName", 50, "testName", "1971", "male", "password");
	@InjectMocks
	RegisterController controller;
	
	@Mock
	PersonRepo repo;
	
	@BeforeEach
	public void setup() {
		MockitoAnnotations.initMocks(this);
	}
	
	@Test
	@DisplayName("save to db-pw lenth>6")
	public void shouldSaveValidPassword1() throws PasswordException {
		//ModelAndView mv = new ModelAndView();
		doReturn(person).when(repo).save(any(Person.class));
		controller.details(person);
		assertEquals(8, person.getPassword().length(), "password length is greater than 6::criteria satisfied");
		verify(repo,times(1)).save(any(Person.class));
	}
}
*/

File: TaskOne/src/test/kotlin/com/ibs/litmus/controllerk/RegisterControllerTest.kt
package com.ibs.litmus.controllerk

import org.mockito.Mockito.doReturn
import com.ibs.litmus.modelk.Person
import com.ibs.litmus.myexceptions.PasswordException
import com.ibs.litmus.repositoryk.PersonRepo
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Assertions.assertThrows
import org.mockito.ArgumentMatchers.any
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations
import org.springframework.web.servlet.mvc.method.annotation.MvcUriComponentsBuilder.on

internal class RegisterControllerTest{
    @InjectMocks
    lateinit var rc : RegisterController
    @Mock
    lateinit var repo : PersonRepo
    lateinit var p : Person

    @BeforeEach
    fun setup(){
        MockitoAnnotations.openMocks(this)
        p = Person("testUserName", "testName", "password", 50, "male", "1971-01-01")
        println("inside setup")
    }

    @AfterEach
    fun cleanup(){println("inside cleanup")}

    @Test
    @DisplayName("save to db:pw len>6")
    @Throws(PasswordException::class)
    fun passwordTest1(){
        p.password = "password"
        //doReturn(p).`when`(repo.save(any(Person::class.java)))
        `when`(repo.save(any(Person::class.java))).thenReturn(p)
        rc.details(p)
        assertEquals(8,p.password.length)
    }

    @Test
    @DisplayName("save to db-pw len=6")
    @Throws(PasswordException::class)
    fun passwordTest2(){
        p.password = "passwo"
        `when`(repo.save(any(Person::class.java))).thenReturn(p)
        rc.details(p)
        assertEquals(6,p.password.length)
    }

    @Test
    @DisplayName("dnt save to db,throw exceptn-pw len<6")
    fun passwordTest3(){
        p.password = "pass"
        assertThrows(Exception::class.java, { -> rc.details(p)},
            "password length is less than 6::criteria violation,throws exception")
    }

}